lib/kv/manager/src/layout/contiguous.rs

```diff
@@ -0,0 +1,528 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! ### Contiguous Memory Layout
+//!
+//! This module provides the implementation for a block-continuous memory layout
+//! of KV blocks.
+//!
+//! A single allocation of memory, minimally of size [nblocks, nlayers, page_size, inner_dim],
+//! is required to
+
+use super::*;
+
+use crate::storage::{Storage, StorageAllocator};
+
+/// Internal struct to hold calculated layout dimensions specific to FullyContiguous.
+// Module-level, but only used internally by FullyContiguous
+#[derive(Debug, Clone)]
+struct FullyContiguousConfig {
+    inner: LayoutConfig,
+    memory_region_size: usize,
+    layer_stride_in_bytes: usize,
+    natural_block_stride: usize,
+    block_stride_in_bytes: usize, // Aligned if necessary
+    layout_data_bytes: usize,     // Size of the layout data itself (post base offset)
+}
+
+impl FullyContiguousConfig {
+    /// Calculates the core dimensions based on the configuration.
+    /// Returns an error if the configuration is invalid.
+    fn new(config: LayoutConfig) -> Result<Self, LayoutError> {
+        // Validate first, propagating errors via `?`
+        config.validate()?;
+
+        let alignment = config.alignment;
+        let memory_region_size = config.page_size * config.inner_dim * config.dtype.size_in_bytes();
+        let layer_stride_in_bytes = memory_region_size;
+        let natural_block_stride = config.num_layers * layer_stride_in_bytes;
+
+        let block_stride_in_bytes = if alignment > 1 {
+            align_up(natural_block_stride, alignment)
+        } else {
+            natural_block_stride
+        };
+
+        let layout_data_bytes =
+            (config.num_blocks - 1) * block_stride_in_bytes + natural_block_stride;
+
+        Ok(Self {
+            inner: config,
+            memory_region_size,
+            layer_stride_in_bytes,
+            natural_block_stride,
+            block_stride_in_bytes,
+            layout_data_bytes,
+        })
+    }
+
+    /// Calculate the total number of bytes required for allocation, including initial alignment padding.
+    /// Panics if the provided configuration is invalid.
+    pub fn required_allocation_size(&self) -> usize {
+        let initial_padding = if self.inner.alignment > 1 {
+            self.inner.alignment - 1
+        } else {
+            0
+        };
+        self.layout_data_bytes + initial_padding
+    }
```

Comment:
https://doc.rust-lang.org/stable/alloc/alloc/struct.Layout.html#method.size

Definitely feels like you're doing a similar thing.